daemon/endpoint.go

```diff
@@ -373,6 +377,13 @@ func (d *Daemon) deleteEndpoint(ep *endpoint.Endpoint) int {
 	for _, err := range errors {
 		scopedLog.WithError(err).Warn("Ignoring error while deleting endpoint")
 	}
+
+	// Since the endpoint is already removed from endpointmanager queue,
+	// we do not need to acquire any lock on it.
+	if len(errors) == 0 {
```

Comment:
This might be even more complicated. `deleteEndpointQuiet` can return early if the endpoint state is disconnecting. I'm not sure if we re-remove it later. In any case, that return happens before the `endpointmanager.Remove` so we would leak disconnecting endpoints.

Reply:
@tgraf : I am trying to understand how it matters if this is called here or in `endpointmanager.Remove()` . I understand the part of not making it conditional with Len(errors), but at this point we would have already called `deleteEndpointQuiet` and removed the endpoint safely from `endpointmanager`. We can simply update the metrics safely at this point? What's the rationale of moving this code to `endpointmanager.Remove() `in `deleteEndpointQuiet()`

Reply:
Additionally we do not need any locking at this point, since the endpoint is no longer in `endpointmanager queue`

Reply:
As the code is as-is, there is no difference but what if endpoints are added and deleted via something other than the API in the future? The endpointmanager controls the lifecycle of an endpoint and the insertion and removal is gated via `endpointmanager.Insert()` and `endpointmanager.Remove()`. Bumping metrics there seems more logical to me.

Reply:
Ah ok..got it! thanks. I will move it there.

Reply:
done.